src/Processors/Transforms/buildPushingToViewsChain.cpp

```diff
@@ -186,6 +186,151 @@ class FinalizingViewsTransform final : public IProcessor
+
+    auto local_select_context = view_metadata_snapshot->getDefinerContext(select_context);
+    auto local_insert_context = view_metadata_snapshot->getDefinerContext(insert_context);
+
+    ASTPtr query;
+    Chain out;
+
+    /// We are creating a ThreadStatus per view to store its metrics individually
+    /// Since calling ThreadStatus() changes current_thread we save it and restore it after the calls
+    /// Later on, before doing any task related to a view, we'll switch to its ThreadStatus, do the work,
+    /// and switch back to the original thread_status.
+    auto * original_thread = current_thread;
+    SCOPE_EXIT({ current_thread = original_thread; });
+    current_thread = nullptr;
+    std::unique_ptr<ThreadStatus> view_thread_status_ptr = std::make_unique<ThreadStatus>(/*check_current_thread_on_destruction=*/ false);
+    /// Copy of a ThreadStatus should be internal.
+    view_thread_status_ptr->setInternalThread();
+    view_thread_status_ptr->attachToGroup(running_group);
+
+    auto * view_thread_status = view_thread_status_ptr.get();
+    views_data->thread_status_holder->thread_statuses.push_front(std::move(view_thread_status_ptr));
+
+    auto runtime_stats = std::make_unique<QueryViewsLogElement::ViewRuntimeStats>();
+    runtime_stats->target_name = view_id.getFullTableName();
+    runtime_stats->thread_status = view_thread_status;
+    runtime_stats->event_time = std::chrono::system_clock::now();
+    runtime_stats->event_status = QueryViewsLogElement::ViewStatus::EXCEPTION_BEFORE_START;
+
+    auto & type = runtime_stats->type;
+    auto & target_name = runtime_stats->target_name;
+    auto * view_counter_ms = &runtime_stats->elapsed_ms;
+
+    if (auto * materialized_view = dynamic_cast<StorageMaterializedView *>(view.get()))
+    {
+        auto lock = materialized_view->tryLockForShare(context->getInitialQueryId(), context->getSettingsRef().lock_acquire_timeout);
+
+        if (lock == nullptr)
+        {
+            // In case the materialized view is dropped/detached at this point, we register a warning and ignore it
+            assert(materialized_view->is_dropped || materialized_view->is_detached);
+            LOG_WARNING(
+                &Poco::Logger::get("PushingToViews"), "Trying to access table {} but it doesn't exist", view_id.getFullTableName());
+            return std::nullopt;
+        }
+
+        type = QueryViewsLogElement::ViewType::MATERIALIZED;
+        result_chain.addTableLock(lock);
+
+        StoragePtr inner_table = materialized_view->getTargetTable();
+        auto inner_table_id = inner_table->getStorageID();
+        auto inner_metadata_snapshot = inner_table->getInMemoryMetadataPtr();
+        query = view_metadata_snapshot->getSelectQuery().inner_query;
+        target_name = inner_table_id.getFullTableName();
+
+        auto header = getHeader(local_select_context, query);
+
+        /// Insert only columns returned by select.
```

Comment:
How about that idea that columns should be ideally filtered if there is a row policy for `local_select_context`?

Reply:
This was copy-pasted from the previous version of `buildPushinToViewsChain` (https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/Transforms/buildPushingToViewsChain.cpp#L346) and isn't fully related to this PR. Probably we can fix it in follow-up PR.